### PR TITLE
Mesh status page

### PR DESCRIPTION
**Pull request by jml**
@bboreham pointed me at the status information for mesh. This has already proved useful in debugging #334, so I figure we might as well get it into master now.

### CODE REVIEW

File: alertmanager/multitenant.go [hunk 1]
```diff
@@ -21,12 +22,55 @@ import (
...
 	minBackoff = 100 * time.Millisecond
 	maxBackoff = 2 * time.Second
+
+	statusPage = `
+<!doctype html>
+<html>
+	<head><title>Cortex Alertmanager Status</title></head>
+	<body>
+		<h1>Cortex Alertmanager Status</h1>
+        <h2>Mesh router</h2>
```
Comment (juliusv) [inline alertmanager/multitenant.go]: spaces vs. tabs, funny indentation here.

File: alertmanager/multitenant.go [hunk 2]
```diff
@@ -21,12 +22,55 @@ import (
...
+        <dl>
+          <dt>Protocol</dt>
+          <dd>{{.Protocol}}
+{{if eq .ProtocolMinVersion .ProtocolMaxVersion}}
+{{.ProtocolMaxVersion}}
+{{else}}
+{{.ProtocolMinVersion}}..{{.ProtocolMaxVersion}}
+{{end}}
```
Comment (juliusv) [inline alertmanager/multitenant.go]: I'd indent this block as well. If you're worried about extra whitespace, you can gobble it up via `-}}` and `{{-`. Quote from docs:

"However, to aid in formatting template source code, if an action's left delimiter (by default "{{") is followed immediately by a minus sign and ASCII space character ("{{- "), all trailing white space is trimmed from the immediately preceding text. Similarly, if the right delimiter ("}}") is preceded by a space and minus sign (" -}}"), all leading white space is trimmed from the immediately following text."

Comment (juliusv) [inline alertmanager/multitenant.go]: Similar for some of the stuff below...

File: alertmanager/multitenant.go [hunk 3]
```diff
@@ -41,11 +85,70 @@ var (
...
 	prometheus.MustRegister(totalConfigs)
+	statusTemplate = template.Must(template.New("statusPage").Funcs(map[string]interface{}{
+		"printState": func(enabled bool) string {
+			if enabled {
+				return "enabled"
+			}
+			return "disabled"
+		},
+		"printConnectionCounts": func(conns []mesh.LocalConnectionStatus) string {
+			counts := make(map[string]int)
```
Comment (juliusv) [inline alertmanager/multitenant.go]: style nit: using `make` only when it's really needed (for chans or slices) seems to have become a more common style in Go projects. So rather `counts := map[string]int{}`.

File: alertmanager/multitenant.go [hunk 4]
```diff
@@ -41,11 +85,70 @@ var (
...
 	}, []string{"operation", "status_code"})
+	statusTemplate      *template.Template
+	allConnectionStates = []string{"established", "pending", "retrying", "failed", "connecting"}
 )
 
 func init() {
 	prometheus.MustRegister(configsRequestDuration)
 	prometheus.MustRegister(totalConfigs)
+	statusTemplate = template.Must(template.New("statusPage").Funcs(map[string]interface{}{
+		"printState": func(enabled bool) string {
```
Comment (juliusv) [inline alertmanager/multitenant.go]: I would probably omit the `print` prefix everywhere, as just naming a function by the thing it returns is more common naming in Go. In Prometheus for example: https://github.com/prometheus/prometheus/blob/master/template/template.go#L119-L251

File: alertmanager/multitenant.go [hunk 5]
```diff
@@ -296,3 +399,32 @@ func (am *MultitenantAlertmanager) ServeHTTP(w http.ResponseWriter, req *http.Re
...
+// ServeHTTP serves the status of the alertmanager.
+func (s StatusHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	meshStatus := mesh.NewStatus(s.am.meshRouter.Router)
+	err := statusTemplate.Execute(w, meshStatus)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+func printState(enabled bool) string {
```
Comment (juliusv) [inline alertmanager/multitenant.go]: Is this kind of left over accidentally? The same function is defined in your function map above.

File: cmd/alertmanager/main.go [hunk 6]
```diff
@@ -51,6 +51,7 @@ func main() {
 	}
 	defer server.Shutdown()
 
+	server.HTTP.PathPrefix("/alertmanager/status").Handler(multiAM.GetStatusHandler())
```
Comment (juliusv) [inline cmd/alertmanager/main.go]: Is the extra `/alertmanager` prefix needed here? We don't have a colliding status page yet, right?

**Comment by juliusv** [comment]
:+1: besides comments (and feel free to ignore the ones you don't like)

File: alertmanager/multitenant.go (hunk 1 above)
Comment (jml) [inline alertmanager/multitenant.go]: Copy/paste from writing the template in an HTML buffer. Go & Make are the only languages where I actually use tab literals.

File: cmd/alertmanager/main.go (hunk 6 above)
Comment (jml) [inline cmd/alertmanager/main.go]: Probably not. I got a bit confused as to how we route this for the cortex ring.

**Pull request merged**